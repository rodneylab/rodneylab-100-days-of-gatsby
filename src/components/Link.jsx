import React from 'react';
import { ExternalLinkIcon } from '@chakra-ui/icons';
import { Link } from '@chakra-ui/react';
import PropTypes from 'prop-types';

export const ExternalLink = ({
  ariaLabel, to, targetBlank, text,
}) => (
  <>
    {targetBlank ? (
      <Link
        aria-label={ariaLabel}
        href={to}
        target="_blank"
        rel="nofollow noopener noreferrer"
        isExternal
      >
        {text}
        {' '}
        <ExternalLinkIcon />
      </Link>
    ) : (
      <Link
        aria-label="Open Rodney Lab contact page"
        href={to}
        rel="noopener"
        isExternal={false}
      >
        {text}
        {' '}
        <ExternalLinkIcon />
      </Link>
    )}
  </>
);

ExternalLink.defaultProps = {
  targetBlank: true,
};

ExternalLink.propTypes = {
  ariaLabel: PropTypes.string.isRequired,
  to: PropTypes.string.isRequired,
  targetBlank: PropTypes.bool,
  text: PropTypes.string.isRequired,
};

export { ExternalLink as default };
