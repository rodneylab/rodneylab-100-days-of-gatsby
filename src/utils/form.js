import { createClient } from '@formium/client';

export const formium = createClient(process.env.GATSBY_FORMIUM_PROJECTID);

export const validate = (values) => {
  const errors = {};
  if (!values.name) {
    errors.name = 'Please enter your name.';
  }
  if (!values.email) {
    errors.email = 'Please enter your email address.';
  } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)) {
    errors.email = 'Please check your email address.';
  }
  if (!values.message) {
    errors.message = 'Please enter a message.';
  } else if (values.message.length > 1024) {
    errors.message = 'Please limit your message to 1000 characters.';
  }
  return errors;
};
