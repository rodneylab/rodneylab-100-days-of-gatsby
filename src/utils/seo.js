import { getSrc } from 'gatsby-plugin-image';

export const getFeaturedImage = ({ featuredImage }) => ({
  url: getSrc(featuredImage.localFile),
  width: 992,
  height: 730,
});

export const getOgImage = ({ ogImage }) => ({
  url: ogImage ? getSrc(ogImage.localFile) : null,
  width: 1200,
  height: 627,
});

export const getOgSquareImage = ({ ogSquareImage }) => ({
  url: ogSquareImage ? getSrc(ogSquareImage.localFile) : null,
  width: 400,
  height: 400,
});

export const getTwitterImage = ({ twitterImage }) => ({
  url: twitterImage ? getSrc(twitterImage.localFile) : null,
  width: 800,
  height: 418,
});
