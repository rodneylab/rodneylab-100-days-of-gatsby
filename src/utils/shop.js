const formatPrice = (currency, value) => Intl.NumberFormat(undefined, {
  currency,
  minimumFractionDigits: 2,
  style: 'currency',
}).format(value);

export { formatPrice as default };

export const localStorageKey = 'a856f4';
