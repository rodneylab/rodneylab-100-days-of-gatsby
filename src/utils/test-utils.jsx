import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
import { createMemorySource, createHistory, LocationProvider } from '@reach/router';

const source = createMemorySource('active');
const history = createHistory(source);

const locationProvider = ({ children }) => (
  <LocationProvider history={history}>{children}</LocationProvider>
);

export const customRender = (ui, options) => render(ui, { wrapper: locationProvider, ...options });

export const getMetaByName = (metaName) => {
  const metas = document.getElementsByTagName('meta');
  for (let i = 0; i < metas.length; i += 1) {
    if (metas[i].getAttribute('name') === metaName) {
      return metas[i].getAttribute('content');
    }
  }
  return '';
};

export const getMetaByProperty = (metaProperty) => {
  const metas = document.getElementsByTagName('meta');
  for (let i = 0; i < metas.length; i += 1) {
    if (metas[i].getAttribute('property') === metaProperty) {
      return metas[i].getAttribute('content');
    }
  }
  return '';
};
