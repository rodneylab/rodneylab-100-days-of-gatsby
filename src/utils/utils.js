// eslint-disable-next-line import/prefer-default-export
export const isBrowser = typeof window !== 'undefined';
