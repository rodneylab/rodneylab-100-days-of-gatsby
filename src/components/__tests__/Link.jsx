import React from 'react';
import { screen } from '@testing-library/react';

import { customRender } from '../../utils/test-utils';

import { ExternalLink } from '../Link';

const ariaLabel = 'aria-label text';

describe('External Link', () => {
  it('renders correctly', async () => {
    const elementToRender = (
      <ExternalLink ariaLabel={ariaLabel} to="https://www.test-site.com" text="test site link" />
    );
    const { asFragment } = customRender(elementToRender);
    expect(asFragment(elementToRender)).toMatchSnapshot();
  });

  it('renders correctly with targetBlank false passed in', async () => {
    const elementToRender = (
      <ExternalLink
        ariaLabel={ariaLabel}
        to="https://www.test-site.com"
        text="test site link"
        targetBlank={false}
      />
    );
    const { asFragment } = customRender(elementToRender);
    expect(asFragment(elementToRender)).toMatchSnapshot();
  });

  it('has target attribute set to _blank by default', () => {
    customRender(
      <ExternalLink ariaLabel={ariaLabel} to="https://www.test-site.com" text="test site link" />,
    );
    expect(screen.getByText('test site link').closest('a')).toHaveAttribute('target', '_blank');
  });

  it('has no target attribute set when targetBlank false is passed in', () => {
    customRender(
      <ExternalLink
        ariaLabel={ariaLabel}
        to="https://www.test-site.com"
        text="test site link"
        targetBlank={false}
      />,
    );
    expect(screen.getByText('test site link').closest('a')).not.toHaveAttribute('target', '_blank');
  });

  it('sets noopener rel attribute', () => {
    customRender(
      <ExternalLink ariaLabel={ariaLabel} to="https://www.test-site.com" text="test site link" />,
    );
    expect(
      screen.getByText('test site link').closest('a').getAttribute('rel').includes('noopener'),
    ).toBeTruthy();
  });

  it('includes noreferrer and nofollow in rel attribute when targetBlank is true', () => {
    customRender(
      <ExternalLink ariaLabel={ariaLabel} to="https://www.test-site.com" text="test site link" targetBlank />,
    );
    expect(
      screen.getByText('test site link').closest('a').getAttribute('rel').includes('noreferrer'),
    ).toBeTruthy();
    expect(
      screen.getByText('test site link').closest('a').getAttribute('rel').includes('nofollow'),
    ).toBeTruthy();
  });
});
