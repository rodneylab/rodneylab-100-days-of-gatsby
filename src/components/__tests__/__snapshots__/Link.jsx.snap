// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`External Link renders correctly 1`] = `
<DocumentFragment>
  <a
    aria-label="aria-label text"
    class="chakra-link css-0"
    href="https://www.test-site.com"
    rel="nofollow noopener noreferrer"
    target="_blank"
  >
    test site link 
    <svg
      class="chakra-icon css-onkibi"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <g
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-width="2"
      >
        <path
          d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
        />
        <path
          d="M15 3h6v6"
        />
        <path
          d="M10 14L21 3"
        />
      </g>
    </svg>
  </a>
</DocumentFragment>
`;

exports[`External Link renders correctly with targetBlank false passed in 1`] = `
<DocumentFragment>
  <a
    aria-label="Open Rodney Lab contact page"
    class="chakra-link css-0"
    href="https://www.test-site.com"
    rel="noopener"
  >
    test site link 
    <svg
      class="chakra-icon css-onkibi"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <g
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-width="2"
      >
        <path
          d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
        />
        <path
          d="M15 3h6v6"
        />
        <path
          d="M10 14L21 3"
        />
      </g>
    </svg>
  </a>
</DocumentFragment>
`;
